From 7fab0d53236492cd27e7b55e2a336fcb08932f55 Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megous@megous.com>
Date: Sun, 29 Nov 2020 06:01:41 +0100
Subject: [PATCH 246/351] usb: typec: anx7688: Move snk/src_caps to the struct

So that they can be accessed later from the PWR_OBJ_REQ handler.
Also they can now be possibly read from DT if needed.

Signed-off-by: Ondrej Jirman <megous@megous.com>
---
 drivers/usb/typec/anx7688.c | 38 ++++++++++++++++++++++++-------------
 1 file changed, 25 insertions(+), 13 deletions(-)

diff --git a/drivers/usb/typec/anx7688.c b/drivers/usb/typec/anx7688.c
index 8b4d20f1b3ae..bfb5ea4f065a 100644
--- a/drivers/usb/typec/anx7688.c
+++ b/drivers/usb/typec/anx7688.c
@@ -172,6 +172,12 @@ struct anx7688 {
         struct gpio_desc *gpio_reset;
         struct gpio_desc *gpio_cabledet;
 
+	uint32_t src_caps[8];
+	unsigned n_src_caps;
+
+	uint32_t snk_caps[8];
+	unsigned n_snk_caps;
+
 	unsigned long flags[1];
 
         struct delayed_work work;
@@ -370,7 +376,6 @@ static int anx7688_connect(struct anx7688 *anx7688)
 	struct typec_partner_desc desc = {};
         int ret, i;
         u8 fw[2];
-        u32 pdo;
 	const u8 dp_snk_identity[16] = {
 		0x00, 0x00, 0x00, 0xec,	/* snk_id_hdr */
 		0x00, 0x00, 0x00, 0x00,	/* snk_cert */
@@ -380,6 +385,7 @@ static int anx7688_connect(struct anx7688 *anx7688)
 	const u8 svid[4] = {
 		0x00, 0x00, 0x01, 0xff,
 	};
+        u32 caps[8];
 
         dev_dbg(anx7688->dev, "cable inserted\n");
 
@@ -477,23 +483,17 @@ static int anx7688_connect(struct anx7688 *anx7688)
 	if (ret)
 		goto err_vconoff;
 
-
-	pdo = cpu_to_le32(PDO_FIXED(5000, 500,
-				    PDO_FIXED_DATA_SWAP |
-				    PDO_FIXED_USB_COMM |
-				    PDO_FIXED_DUAL_ROLE));
+	for (i = 0; i < anx7688->n_src_caps; i++)
+		caps[i] = cpu_to_le32(anx7688->src_caps[i]);
 	ret = anx7688_send_ocm_message(anx7688, ANX7688_OCM_MSG_PWR_SRC_CAP,
-				       (u8*)&pdo, sizeof pdo);
+				       (u8*)&caps, 4 * anx7688->n_src_caps);
 	if (ret)
 		goto err_vconoff;
 
-
-	pdo = cpu_to_le32(PDO_FIXED(5000, 3000,
-				    PDO_FIXED_DATA_SWAP |
-				    PDO_FIXED_USB_COMM |
-				    PDO_FIXED_DUAL_ROLE));
+	for (i = 0; i < anx7688->n_snk_caps; i++)
+		caps[i] = cpu_to_le32(anx7688->snk_caps[i]);
 	ret = anx7688_send_ocm_message(anx7688, ANX7688_OCM_MSG_PWR_SNK_CAP,
-				       (u8*)&pdo, sizeof pdo);
+				       (u8*)&caps, 4 * anx7688->n_snk_caps);
 	if (ret)
 		goto err_vconoff;
 
@@ -1534,6 +1534,18 @@ static int anx7688_i2c_probe(struct i2c_client *client,
         mutex_init(&anx7688->lock);
         INIT_DELAYED_WORK(&anx7688->work, anx7688_work);
 
+	anx7688->n_src_caps = 1;
+	anx7688->src_caps[0] = PDO_FIXED(5000, 500,
+					 PDO_FIXED_DATA_SWAP |
+					 PDO_FIXED_USB_COMM |
+					 PDO_FIXED_DUAL_ROLE);
+
+	anx7688->n_snk_caps = 1;
+	anx7688->snk_caps[0] = PDO_FIXED(5000, 3000,
+					 PDO_FIXED_DATA_SWAP |
+					 PDO_FIXED_USB_COMM |
+					 PDO_FIXED_DUAL_ROLE);
+
         for (i = 0; i < ANX7688_NUM_SUPPLIES; i++)
                 anx7688->supplies[i].supply = anx7688_supply_names[i];
         ret = devm_regulator_bulk_get(dev, ANX7688_NUM_SUPPLIES,
-- 
2.34.0

